File: Atividades/2024-03-13/Arvores/arvores_binarias_atividade/src/main/java/dev/mandevilla/Models/NoModel.java
package dev.mandevilla.Models;

public class NoModel {
    private int _valor;
    private NoModel _esquerdo;
    private NoModel _direito;

    public NoModel(int valor) {
        _valor = valor;
        _esquerdo = null;
        _direito = null;
    }

    // _valor : int
    public int getValor() {
        return _valor;
    }

    public void setValor(int _valor) {
        this._valor = _valor;
    }

    // _esquerdo : NoModel
    public NoModel getEsquerdo() {
        return _esquerdo;
    }

    public void setEsquerdo(NoModel esquerdo) {
        this._esquerdo = esquerdo;
    }

    public void setEsquerdo(int valor) {
        this._esquerdo = new NoModel(valor);
    }

    // _direito : NoModel
    public NoModel getDireito() {
        return _direito;
    }

    public void setDireito(NoModel direito) {
        this._direito = direito;
    }

    public void setDireito(int valor) {
        this._direito = new NoModel(valor);
    }
}

File: Atividades/2024-03-27/Arvores/arvores_binarias_simulado/lib/src/shared/models/no/no_model.dart
/// Classe modelo para os nós da Árvore binária.
class NoModel {
  /// Valor do nó.
  final int valor;

  /// Nó a esquerda.
  late NoModel? _esquerdo;

  /// Nó a direita.
  late NoModel? _direito;

  /// Construtor da classe.
  ///
  /// [valor] é obrigatório.
  NoModel({
    required this.valor,
    NoModel? esquerdo,
    NoModel? direito,
  }) {
    _esquerdo = esquerdo;
    _direito = direito;
  }

  /// Retorna o valor referente ao nó a esquerda.
  NoModel? get esquerdo => _esquerdo;

  /// Retorna o valor referente ao nó a direita.
  NoModel? get direito => _direito;

  /// Define o valor referente ao nó a esquerda.
  void setEsquerdo(int valor) {
    _esquerdo = NoModel(valor: valor);
  }

  /// Define o valor referente ao nó a direita.
  void setDireito(int valor) {
    _direito = NoModel(valor: valor);
  }

  /// Retorna ***true*** caso o nó seja **folha**, ou seja, [esquerdo] e [direito] forem iguais a [null].
  ///
  /// Caso contrário, retornará ***false***.
  bool get isFolha {
    return _esquerdo == null && _direito == null;
  }
}
